packages/react-instantsearch/src/components/useAutocomplete.tsx

```diff
@@ -0,0 +1,248 @@
+import { useId, useLayoutEffect, useRef, useState } from 'react';
+import { useInstantSearch } from 'react-instantsearch-core';
+
+import type { ComponentProps, Ref } from 'react';
+
+type AutocompleteGetInputProps = () => Pick<
+  ComponentProps<'input'>,
+  | 'id'
+  | 'role'
+  | 'aria-activedescendant'
+  | 'aria-autocomplete'
+  | 'aria-controls'
+  | 'aria-expanded'
+  | 'aria-haspopup'
+  | 'onBlur'
+  | 'onFocus'
+  | 'onKeyDown'
+  | 'onKeyUp'
+>;
+
+type AutocompleteGetItemProps = (
+  item: {
+    objectID: string;
+    __indexName?: string;
+  } & Record<string, unknown>,
+  index: number
+) => Pick<
+  ComponentProps<'div'>,
+  'id' | 'role' | 'aria-selected' | 'onMouseEnter' | 'onMouseLeave'
+>;
+
+type AutocompleteGetPanelProps = () => Pick<
+  ComponentProps<'div'>,
+  'id' | 'role' | 'aria-labelledby'
+> & {
+  ref: Ref<HTMLElement>;
+  isOpen: boolean;
+};
+
+type AutocompleteGetRootProps = () => {
+  ref: Ref<HTMLDivElement>;
+};
+
+type AutocompleteUpdateStoreParams = {
+  indexName: string;
+  items: Array<{ objectID: string } & Record<string, unknown>>;
+  onSelect: (item: { objectID: string } & Record<string, unknown>) => void;
+  getQuery?: (item: { objectID: string } & Record<string, unknown>) => string;
+  getURL?: (item: { objectID: string } & Record<string, unknown>) => string;
+};
+
+export const useAutocomplete = () => {
+  const { setUiState } = useInstantSearch();
+
+  const getElementId = createGetElementId(useId());
+  const rootRef = useRef<HTMLInputElement>(null);
+  const panelRef = useRef<HTMLDivElement>(null);
+  const [isOpen, setIsOpen] = useState(false);
+  const [activeDescendent, setActiveDescendent] = useState<string | undefined>(
+    undefined
+  );
+
+  const store = new Map<
+    string,
+    {
+      item: AutocompleteUpdateStoreParams['items'][0];
+      getQuery: () => string;
+      getURL: () => string;
+    } & Omit<AutocompleteUpdateStoreParams, 'items' | 'getQuery' | 'getURL'>
+  >();
+
+  useLayoutEffect(() => {
+    const onBodyClick = (event: MouseEvent) => {
+      if (rootRef.current?.contains(event.target as HTMLElement)) {
+        return;
+      }
+
+      setIsOpen(false);
+    };
+
+    document.body.addEventListener('click', onBodyClick);
+
+    return () => {
+      document.body.removeEventListener('click', onBodyClick);
+    };
+  }, [rootRef]);
+
+  const getNextActiveDescendent = (key: string): string | undefined => {
```

Comment:
this function could be declared outside of the bode of useAutocomplete if it received the state as a prop. Not sure if it would be better though.